File: Apps/iOS/Domain/UseCases/PlayScore.swift
import Foundation

final class PlayScore {
    private static let articulationSeconds = 0.05
    private static let slideStepSeconds = 0.04

    private let tuning: RichterTuning
    private let harmonica: PlayHarmonica
    private let log: LogProtocol

    // MARK: - Public

    init(tuning: RichterTuning, harmonica: PlayHarmonica, log: LogProtocol) {
        self.tuning = tuning
        self.harmonica = harmonica
        self.log = log
    }

    func play(_ score: Score) -> AsyncStream<Harmonica> {
        AsyncStream { continuation in
            let performance = Task { await perform(score, into: continuation) }
            continuation.onTermination = { _ in performance.cancel() }
        }
    }

    // MARK: - Private

    private func perform(_ score: Score, into continuation: AsyncStream<Harmonica>.Continuation) async {
        log.record("score started in \(score.key), \(score.position) position, calling for a harmonica in \(score.harmonicaKey)")
        continuation.yield(harmonica.changeKey(to: score.harmonicaKey))
        for event in score.events where !Task.isCancelled {
            await perform(event, secondsPerBeat: score.secondsPerBeat, into: continuation)
        }
        continuation.yield(harmonica.shapeTone(.rest, vibrato: .off))
        continuation.yield(harmonica.stopPlaying())
        log.record(Task.isCancelled ? "score stopped early" : "score finished")
        continuation.finish()
    }

    private func perform(
        _ event: ScoreEvent,
        secondsPerBeat: Double,
        into continuation: AsyncStream<Harmonica>.Continuation
    ) async {
        let seconds = event.beats * secondsPerBeat
        guard case .note(let note) = event else {
            continuation.yield(harmonica.stopPlaying())
            await wait(seconds)
            return
        }

        let slid = await slide(into: note, within: seconds, into: continuation)
        await sound(note, for: seconds - slid, into: continuation)
    }

    private func slide(
        into note: ScoreNote,
        within seconds: Double,
        into continuation: AsyncStream<Harmonica>.Continuation
    ) async -> Double {
        let passing = Self.passingHoles(of: note)
        guard !passing.isEmpty else { return 0 }

        let step = min(Self.slideStepSeconds, seconds / 2 / Double(passing.count))
        continuation.yield(harmonica.shapeTone(.rest, vibrato: .off))
        for hole in passing {
            continuation.yield(harmonica.play(at: [Self.position(of: hole, breathing: note.breath)]))
            await wait(step)
        }
        return step * Double(passing.count)
    }

    private func sound(
        _ note: ScoreNote,
        for seconds: Double,
        into continuation: AsyncStream<Harmonica>.Continuation
    ) async {
        let gap = min(Self.articulationSeconds, seconds / 4)
        continuation.yield(harmonica.shapeTone(shaping(for: note), vibrato: VibratoDepth(clamping: note.vibrato)))
        continuation.yield(harmonica.play(at: Self.positions(of: note)))
        await wait(seconds - gap)
        continuation.yield(harmonica.stopPlaying())
        await wait(gap)
    }

    private func wait(_ seconds: Double) async {
        try? await Task.sleep(for: .seconds(max(0, seconds)))
    }

    private func shaping(for note: ScoreNote) -> PitchShaping {
        guard !note.isOverbent else { return PitchShaping(clamping: 1) }

        let range = note.holes
            .map { tuning.bendableSemitones(for: Reed(hole: $0, breath: note.breath)) }
            .max() ?? 0
        guard note.bentBySemitones > 0, range > 0 else { return .rest }

        return PitchShaping(clamping: -note.bentBySemitones / range)
    }

    private static func passingHoles(of note: ScoreNote) -> [Hole] {
        guard let from = note.slideFrom, let arriving = note.holes.first, from != arriving else { return [] }

        let numbers = from.number < arriving.number
            ? Array(from.number..<arriving.number)
            : Array((arriving.number + 1...from.number).reversed())
        return numbers.compactMap(Hole.init(rawValue:))
    }

    private static func positions(of note: ScoreNote) -> [PositionOnHarmonica] {
        note.holes.map { position(of: $0, breathing: note.breath) }
    }

    private static func position(of hole: Hole, breathing breath: Breath) -> PositionOnHarmonica {
        PositionOnHarmonica(
            fractionFromLeftEdge: (Double(hole.number) - 0.5) / Double(Hole.allCases.count),
            fractionAboveCentreLine: breath == .blow ? 0.5 : -0.5
        )
    }
}

File: Web/src/domain/useCases/playScore.ts
import type { Harmonica } from '../instrument/harmonica.js'
import type { Hole } from '../instrument/hole.js'
import { holes as everyHole } from '../instrument/hole.js'
import { bendableSemitones } from '../instrument/richterTuning.js'
import type { PitchShaping } from '../playing/pitchShaping.js'
import { pitchShaping, restingShaping } from '../playing/pitchShaping.js'
import { vibratoDepth, vibratoOff } from '../playing/vibratoDepth.js'
import type { Log } from '../protocols/log.js'
import type { Score, ScoreEvent, ScoreNote } from '../scores/score.js'
import { harmonicaKeyOf, secondsPerBeat } from '../scores/score.js'
import type { PlayHarmonica } from './playHarmonica.js'

const articulationSeconds = 0.05
const slideStepSeconds = 0.04
const shakeStepSeconds = 0.06
const bendStepSeconds = 0.01

export interface ScorePerformance {
    readonly finished: Promise<void>
    cancel(): void
}

interface Cancellation {
    stopped: boolean
}

export class PlayScore {
    constructor(
        private readonly harmonica: PlayHarmonica,
        private readonly log: Log
    ) {}

    play(score: Score, played: (harmonica: Harmonica) => void): ScorePerformance {
        const cancellation: Cancellation = { stopped: false }
        const finished = this.perform(score, played, cancellation)
        return { finished, cancel: () => { cancellation.stopped = true } }
    }

    private async perform(
        score: Score,
        played: (harmonica: Harmonica) => void,
        cancellation: Cancellation
    ): Promise<void> {
        this.log.record(
            `score started in ${score.key}, ${score.position} position,`
            + ` calling for a harmonica in ${harmonicaKeyOf(score)}`
        )
        played(this.harmonica.changeKey(harmonicaKeyOf(score)))
        for (const event of score.events) {
            if (cancellation.stopped) break

            await this.performEvent(event, secondsPerBeat(score), played, cancellation)
        }
        played(this.harmonica.shapeTone(restingShaping, vibratoOff))
        played(this.harmonica.stopPlaying('ringsDown'))
        this.log.record(cancellation.stopped ? 'score stopped early' : 'score finished')
    }

    private async performEvent(
        event: ScoreEvent,
        secondsPerBeat: number,
        played: (harmonica: Harmonica) => void,
        cancellation: Cancellation
    ): Promise<void> {
        const seconds = (event.kind === 'note' ? event.note.beats : event.beats) * secondsPerBeat
        if (event.kind === 'rest') return wait(seconds)

        const slid = await this.slide(event.note, seconds, played, cancellation)
        await this.sound(event.note, seconds - slid, played, cancellation)
    }

    private async slide(
        note: ScoreNote,
        seconds: number,
        played: (harmonica: Harmonica) => void,
        cancellation: Cancellation
    ): Promise<number> {
        const passing = passingHoles(note)
        if (passing.length === 0) return 0

        const step = Math.min(slideStepSeconds, seconds / 2 / passing.length)
        played(this.harmonica.shapeTone(restingShaping, vibratoOff))
        for (const hole of passing) {
            if (cancellation.stopped) break

            played(this.harmonica.play([hole], note.breath))
            await wait(step)
        }
        return step * passing.length
    }

    private async sound(
        note: ScoreNote,
        seconds: number,
        played: (harmonica: Harmonica) => void,
        cancellation: Cancellation
    ): Promise<void> {
        const gap = Math.min(articulationSeconds, seconds / 4)
        const sounding = seconds - gap
        const steps = stepsOf(note, sounding)
        const started = now()
        for (let step = 0; step < steps; step += 1) {
            if (cancellation.stopped) break

            this.harmonica.shapeTone(
                this.shapingFor(note, fractionOf(step, steps)),
                vibratoDepth(note.vibrato ?? 0)
            )
            played(this.harmonica.play(holesOf(note, step), note.breath))
            await waitUntil(started + (sounding * (step + 1)) / steps)
        }
        played(this.harmonica.stopPlaying('damped'))
        await wait(gap)
    }

    private shapingFor(note: ScoreNote, fraction: number): PitchShaping {
        if (note.isOverbent === true) return pitchShaping(1)

        const semitones = bendOf(note, fraction)
        const range = Math.max(
            0,
            ...note.holes.map(hole => bendableSemitones({ hole, breath: note.breath }))
        )
        if (semitones <= 0 || range <= 0) return restingShaping

        return pitchShaping(-semitones / range)
    }
}

function stepsOf(note: ScoreNote, seconds: number): number {
    const step = stepSecondsOf(note)
    if (step === null) return 1

    return Math.max(1, Math.floor(seconds / step))
}

function stepSecondsOf(note: ScoreNote): number | null {
    if (note.shakenWith !== undefined) return shakeStepSeconds

    return note.bendEndsAtSemitones === undefined ? null : bendStepSeconds
}

function fractionOf(step: number, steps: number): number {
    return steps > 1 ? step / (steps - 1) : 0
}

function bendOf(note: ScoreNote, fraction: number): number {
    const bent = note.bentBySemitones ?? 0
    const ending = note.bendEndsAtSemitones
    if (ending === undefined) return bent

    return bent + (ending - bent) * fraction
}

function passingHoles(note: ScoreNote): Hole[] {
    const from = note.slideFrom
    const arriving = note.holes[0]
    if (from === undefined || arriving === undefined || from === arriving) return []

    const numbers = from < arriving
        ? everyHole.filter(hole => hole >= from && hole < arriving)
        : everyHole.filter(hole => hole > arriving && hole <= from).reverse()
    return numbers
}

function holesOf(note: ScoreNote, step: number): readonly Hole[] {
    const shaken = note.shakenWith
    if (shaken === undefined || step % 2 === 0) return note.holes

    return [shaken]
}

function now(): number {
    return performance.now() / 1000
}

function wait(seconds: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, Math.max(0, seconds) * 1000))
}

function waitUntil(seconds: number): Promise<void> {
    return wait(seconds - now())
}
